crates/autopilot/src/domain/events/mod.rs

```diff
@@ -0,0 +1,205 @@
+use {
+    crate::{
+        boundary,
+        domain::{self, eth},
+        infra::{
+            self,
+            persistence::{self, transaction::Transaction},
+        },
+    },
+    anyhow::Context,
+    ethrpc::current_block::RangeInclusive,
+    std::sync::Arc,
+};
+
+pub struct Events {
+    persistence: Arc<infra::Persistence>,
+}
+
+/// Error type for the events module.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("persistence error: {0:?}")]
+    Persistence(#[from] persistence::Error),
+    #[error("unexpected event data: {0:?}")]
+    Encoding(anyhow::Error),
+}
+
+pub mod settlement {
+    use super::*;
+
+    /// An order was fully or partially traded.
+    pub struct Trade {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub uid: domain::OrderUid,
+        pub sell_amount_including_fee: eth::U256,
+        pub buy_amount: eth::U256,
+        pub fee_amount: eth::U256,
+    }
+
+    /// An order was cancelled on-chain.
+    pub struct Cancellation {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub uid: domain::OrderUid,
+    }
+
+    /// A settlement was executed on-chain.
+    pub struct Settlement {
+        pub block_number: u64,
+        /// The index of the event in the block.
+        pub log_index: usize,
+        pub tx_hash: eth::H256,
+        pub solver: eth::Address,
+    }
+
+    /// An order was signed on-chain or signature for an on-chain signed order
+    /// has been revoked.
+    pub struct PreSignature {
```

Comment:
`PreSignature` and `Cancellation` should arguably not be part of the `settlement` module. `events` seems more fitting.

Reply:
The module is supposed to capture all events from the GPv2Settlement contract instance (of which PreSignature and Cancellation are a part of). We also have the ETHFlow contract (which is currently handled by two separate indexers), which I'd ideally like to integrate here as well.

Maybe making the module `domain::events::contract::settlement` makes this clearer?

Reply:
Was about to update my initial commit because I realized that it refers to all the settlement contract events. :sweat_smile:
I think throwing the in a `contracts` module might help with that.